src/Psl/Regex/matching.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Regex;
+
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Psl\Type\TypeInterface;
+
+use function preg_match;
+
+/**
+ * Determine if $subject matches the given $pattern and return the matches.
+ *
+ * @template T of array
+ *
+ * @param non-empty-string $pattern        The pattern to match against.
+ * @param TypeInterface<T> $capture_groups What shape does the matching items have?
+ *
+ * @return T|null
+ *
+ * @throws Exception\RuntimeException        If an internal error accord.
+ * @throws Exception\InvalidPatternException If $pattern is invalid.
+ */
+function matching(string $subject, string $pattern, TypeInterface $capture_groups, int $offset = 0): ?array
+{
+    $matching = Internal\call_preg(
+        'preg_match',
+        static function () use ($subject, $pattern, $offset): ?array {
+            $matching = [];
+            $matches = preg_match($pattern, $subject, $matching, 0, $offset);
+
+            return $matches === 0 ? null : $matching;
+        }
+    );
+
+    if ($matching === null) {
+        return null;
+    }
+
+    try {
+        return $capture_groups->assert($matching);
+    } catch (AssertException $e) {
+    }
+
+    try {
+        return $capture_groups->coerce($matching);
+    } catch (CoercionException $e) {
+        return null;
```

Comment:
What do you want it to do if the coercing fails?
Throw a runtime exception, null or the matching results?
I'dd say throw

Reply:
throw runtime exception, and we set the coercion exception as the previous exception

`throw new Exception\RuntimeException('Invalid capture groups', 0, $e);`

Reply:
Introduced a new `Type\cast() function to avoid some double try-catches in the code.`